html/user/job_file.php

```diff
@@ -162,30 +162,49 @@ function upload_files($r) {
     $delete_time = (int)$r->delete_time;
     $batch_id = (int)$r->batch_id;
     //print_r($_FILES);
-    $i = 0;
+    $i = -1;
+    $upload_error = false;
+    $move_error = false;
+    $db_error = false;
     foreach ($r->md5 as $f) {
+        $i++;
         $md5 = (string)$f;
         $name = "file_$i";
         $tmp_name = $_FILES[$name]['tmp_name'];
         if (!is_uploaded_file($tmp_name)) {
-            xml_error(-1, "$tmp_name is not an uploaded file");
+            $upload_error = "$name:$md5 was not uploaded";
+            continue;
         }
         $fname = job_file_name($md5);
         $path = dir_hier_path($fname, project_dir() . "/download", $fanout);
-        rename($tmp_name, $path);
+        if (!rename($tmp_name, $path)) {
+            $move_error = "could not move $tmp_name to $path";
+            unlink($tmp_name);
+            continue;
+        }
         $now = time();
         $jf_id = BoincJobFile::insert(
             "(md5, create_time, delete_time) values ('$md5', $now, $delete_time)"
         );
         if (!$jf_id) {
-            xml_error(-1, "upload_files(): BoincJobFile::insert($md5) failed: ".BoincDb::error());
+            $db_error = "BoincJobFile::insert($md5) failed: ".BoincDb::error();
+            $unlink($path);
+            continue;
         }
         if ($batch_id) {
             BoincBatchFileAssoc::insert(
                 "(batch_id, job_file_id) values ($batch_id, $jf_id)"
             );
         }
-        $i++;
+    }
+    if ($move_error) {
```

Comment:
"this" refers to the three if statements on line 200ff. 

Use-Case: uploading two files (file1.zip, file2.zip) which both could not be moved to the download directory. Because of line 181 which is called once for each file, $move_error at the end (on line 200) only contains "could not move /tmp/file2.zip to download/*/file2.zip" thus hiding the error with file1.zip

A quick fix would be to initialize `$*_error` with an empty string. Concatenating the error messages and checking for `strlen($*_error)>0` in the if blocks at the end.

To catch all errors one could also only use one `$error_str` variable and concatenate the errors to it and call xml_error() only once reporting all the errors that happened.